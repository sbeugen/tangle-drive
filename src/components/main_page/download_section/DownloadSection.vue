<template>
  <div>
    <p>1. Simply paste your bundle hash and download the file.</p>
    <input type="text" class="bundle-hash" v-model="bundleHash" placeholder="Paste your bundle hash">
    <br>
    <br>
    <template v-if="bundleHash">
      <p>2. Download the file.</p>
      <button @click="downloadClickHandler" :disabled="disabled">Download</button>
    </template>
  </div>
</template>

<script>
  import { mapActions, mapGetters } from 'vuex'; 
  import iota from '@/plugins/iota' 

  export default {
    data() {
      return {
        bundleHash: '',
        disabled: false
      }
    },
    methods: {
      async downloadClickHandler() {
        if (this.bundleHash.length != 81) {
          alert('The Bundle Hash has to be exactly 81 characters long!')
        } else {
          try {
            this.disabled = true

            await this.prepareDownload(this.bundleHash)

            let a = document.createElement('a')
            document.body.appendChild(a)
            a.style = "display: none"
            a.href = this.getFileURL
            a.download = this.getFileName
            a.click()
            window.URL.revokeObjectURL(this.getFileURL)
          } catch(error) {
            alert(error)
          }
          this.disabled = false
        }
      },
      ...mapActions('download', [
        'prepareDownload'
      ])
    },
    computed: {
      ...mapGetters('download', [
        'getFileURL',
        'getFileName'
      ])
    }
  }
  
</script>

<style scoped>
  button {
    border: 1px solid white;
    background: skyblue;
    padding: 10px 24px; 
    cursor: pointer;
    margin: 0;
  }
  button:hover {
    border-color: skyblue;
  }
  .bundle-hash {
    width: 750px;
    padding: 9px 5px;
  }
</style>
