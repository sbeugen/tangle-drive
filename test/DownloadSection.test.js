import { shallow, createLocalVue } from 'vue-test-utils'
import Vuex from 'vuex'
import Vue from 'vue'

import DownloadSection from '../src/components/main_page/download_section/DownloadSection'

const localVue = createLocalVue()
localVue.use(Vuex)

const BUNDLE_HASH = '123456789012345678901234567890123456789012345678901234567890123456789012345678901'

describe('DownloadSection.test.js', () => { 
  let store
  let uploadActions
  let downloadActions
  let downloadGetters
  let cmp

  uploadActions = {
    setUploadActive: jest.fn()
  }
  downloadActions = {
    setDownloadActive: jest.fn(),
    prepareDownload: jest.fn()
  }
  downloadGetters = {
    getFileURL : jest.fn().mockImplementation(() => ''),
    getFileName : jest.fn().mockImplementation(() => '')
  }

  beforeEach(() => {
    store = new Vuex.Store({
      strict: true,
      modules: {
        upload: {
          namespaced: true,
          actions: uploadActions
        },
        download: {
          namespaced: true,
          getters: downloadGetters,
          actions: downloadActions
        }
      }
    })

    cmp = shallow(DownloadSection, { store, localVue })
  })

  describe('Directly after mount', () => {
    it('has expected html structure', () => {
      expect(cmp.element).toMatchSnapshot()
    })

    it('Default values are set correctly', () => {
      expect(cmp.vm.disabled).toEqual(false)
      expect(cmp.vm.bundleHash).toEqual('')
    })

    it('functions inside beforeMount are called', () => {
      expect(uploadActions.setUploadActive).toBeCalled()
      expect(downloadActions.setDownloadActive).toBeCalled()
    })
  })

  describe('When something was filled in in bundleHashInput', () => {
    it('bundleInputChangeHandler is called', () => {
      cmp.vm.bundleInputChangeHandler = jest.fn()
      cmp.find('input').trigger('change')
      expect(cmp.vm.bundleInputChangeHandler).toBeCalled()
    })

    it('Two p-elements and one button are visible', () => {
      cmp.setData({ bundleHash: BUNDLE_HASH })
      expect(cmp.findAll('p').length).toEqual(2)
      expect(cmp.findAll('button').length).toEqual(1)
    })
  })

  describe('When downloadButton is clicked', () => {
    it('downloadClickHandler is called', () => {
      cmp.setData({ bundleHash: BUNDLE_HASH })
      cmp.vm.downloadClickHandler = jest.fn()
      cmp.find('button').trigger('click')
      expect(cmp.vm.downloadClickHandler).toBeCalled()
    })

    it('prepareDownload action is called inside downloadClickHandler', () => {
      window.URL.revokeObjectURL = jest.fn()
      cmp.setData({ bundleHash: BUNDLE_HASH })
      cmp.find('button').trigger('click')
      expect(downloadActions.prepareDownload).toBeCalled()
    })
  })
})