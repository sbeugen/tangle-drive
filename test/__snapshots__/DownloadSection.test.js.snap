// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadSection.test.js Directly after mount has expected html structure 1`] = `
<div>
  <p>
    1. Simply paste your bundle hash and download the file.
  </p>
   
  <input
    class="bundle-hash"
    placeholder="Paste your bundle hash"
    type="text"
  />
   
  <br />
   
  <br />
   
  <!---->
</div>
`;
